timemachine/fe/absolute_hydration.py

```diff
@@ -0,0 +1,115 @@
+"""Absolute hydration free energies"""
+
+from functools import partial
+
+import numpy as np
+from tqdm import tqdm
+
+from timemachine.constants import BOLTZ
+from timemachine.fe import free_energy
+from timemachine.fe.free_energy_rabfe import construct_pre_optimized_absolute_lambda_schedule_solvent
+from timemachine.md import builders, enhanced, moves
+from timemachine.md.smc import conditional_multinomial_resample
+from timemachine.md.states import CoordsVelBox
+from timemachine.utils import bind_potentials, construct_potential, get_ff_am1ccc
+
+
+def generate_endstate_samples(num_samples, solvent_samples, ligand_samples, ligand_log_weights, num_ligand_atoms):
+    """solvent + (noninteracting ligand) sample --> solvent + (vacuum ligand) sample
+
+    Assumptions:
+    ------------
+    * ligand indices: last num_ligand_atoms
+
+    TODO: document me more"""
+    all_xvbs = []
+    for _ in tqdm(range(num_samples), desc="generating endstate samples"):
+        choice_idx = np.random.choice(np.arange(len(solvent_samples)))
+        solvent_x = solvent_samples[choice_idx].coords
+        solvent_v = solvent_samples[choice_idx].velocities
+        ligand_xv = enhanced.sample_from_log_weights(ligand_samples, ligand_log_weights, size=1)[0]
+        ligand_x = ligand_xv[0]
+        ligand_v = ligand_xv[1]
+        combined_x = np.concatenate([solvent_x[:-num_ligand_atoms], ligand_x], axis=0)
+        combined_v = np.concatenate([solvent_v[:-num_ligand_atoms], ligand_v], axis=0)
+        combined_box = solvent_samples[choice_idx].box
+        all_xvbs.append(CoordsVelBox(combined_x, combined_v, combined_box))
+    return all_xvbs
+
+
+def get_solvent_phase_system(mol, ff):
+    water_system, water_coords, water_box, water_topology = builders.build_water_system(3.0)
+    water_box = water_box + np.eye(3) * 0.5  # add a small margin around the box for stability
+    afe = free_energy.AbsoluteFreeEnergy(mol, ff)
+    ff_params = ff.get_ordered_params()
+    ubps, params, masses, coords = afe.prepare_host_edge(ff_params, water_system, water_coords)
+    return ubps, params, masses, coords, water_box
```

Comment:
Can probably discard this one -- superseded by implementation in md/enhanced.py -- thanks to @mcwitt for taking a look!

Reply:
Applied in https://github.com/proteneer/timemachine/pull/718/commits/65dbef2dc35caac29dc2be3d2993d97dd36d6fd6